src/sparseml/pytorch/optim/manager.py

```diff
@@ -87,67 +265,147 @@ def from_yaml(
+        epoch and steps_per_epoch can optionally be passed in to initialize the manager
+        and module at a specific point in the training process.
+        If loggers is not None, will additionally call initialize_loggers.
+
+        :param module: the PyTorch model/module to modify
+        :param epoch: The epoch to initialize the manager and module at.
+            Defaults to 0 (start of the training process)
+        :param loggers: Optional list of loggers to log the modification process to
+        :param kwargs: Optional kwargs to support specific arguments
+            for individual modifiers.
+        """
+        super().initialize(module, epoch, loggers, **kwargs)
+        self._initialize_epoch = epoch
+
+        for mod in self._modifiers:
+            mod.initialize(module, epoch, loggers, **kwargs)
 
     def initialize_loggers(self, loggers: Union[None, List[BaseLogger]]):
         """
-        Handles initializing and setting up the loggers for the contained modifiers
-        Called once on construction of the scheduled optimizer
+        Handles initializing and setting up the loggers for the contained modifiers.
 
-        :param loggers: the loggers to setup this modifier with for logging important
+        :param loggers: the loggers to setup this manager with for logging important
             info and milestones to
         """
         super().initialize_loggers(loggers)
 
         for mod in self._modifiers:
             mod.initialize_loggers(loggers)
 
+    def modify(
+        self,
+        module: Module,
+        optimizer: Optimizer,
+        steps_per_epoch: int,
+        wrap_optim: Any = None,
+        epoch: float = None,
+    ) -> ModifierManagerOptimWrapper:
+        """
+        Modify the given module and optimizer for training aware algorithms such as
+        pruning and quantization.
+        Initialize must be called first.
+        After training is complete, finalize should be called.
+
+        :param module: The model/module to modify
+        :param optimizer: The optimizer to modify
+        :param steps_per_epoch: The number of optimizer steps (batches) in each epoch
+        :param wrap_optim: Optional object to wrap instead of the optimizer.
+            Useful for cases like amp (fp16 training) where a it should be wrapped
+            in place of the original optimizer since it doesn't always call into
+            the optimizer.step() function.
+        :param epoch: Optional epoch that can be passed in to start modifying at.
+            Defaults to the epoch that was supplied to the initialize function.
+        :return: A wrapped optimizer object. The wrapped object makes all the
+            original properties for the wrapped object available so it can be
+            used without any additional code changes.
+        """
+        if not self.initialized:
+            raise RuntimeError("manager must be initialized first")
```

Comment:
can we call into initialize instead of failing here? an implicit call to initialize would save an extra line in standard integrations.

Reply:
yes, will default to calling into initialize if not called. Will also work out well for TF and keras